store/cachekv/store.go

```diff
@@ -49,21 +31,34 @@ func (store *Store) GetStoreType() types.StoreType {
 	return store.parent.GetStoreType()
 }
 
+// Clone creates a snapshot of the cache store.
+// This is a copy-on-write operation and is very fast because
+// it only performs a shadowed copy.
+func (store *Store) Clone() types.CacheKVStore {
+	return &Store{
+		cache:  store.cache.Copy(),
+		parent: store.parent,
+	}
+}
+
+// Restore restores the store cache to a given snapshot.
+func (store *Store) Restore(s types.CacheKVStore) {
+	ss := s.(*Store)
+	store.cache = ss.cache
```

Comment:
You can't assume the `s` parameter is a pointer to a concrete Store as defined in this package, it can be any concrete type that satisfies the CacheKVStore interface, which this code must accommodate.

You can't access fields of that store value without first acquiring the store mutex, otherwise you have a data race and memory model violation.

Assuming the received value is actually a concrete Store type, its cache field is an interface, and so has reference semantics. Copying it in this way means the receiver `*Store` and the origin `s` will both have the same pointer to the same value. Both of those stores will be able to access the underlying store concurrently. This is a data race as well.

Reply:
> good catch, I should clear the original store.

A name like "Restore" doesn't suggest that the parameter you give to it will be erased. You want to copy the original store into the method receiver, rather than do a destructive move, I think.